core-api/src/main/java/com/optimizely/ab/config/audience/match/AttributeMatch.java

```diff
@@ -21,8 +21,8 @@ T castToValueType(Object o, Object value) {
         try {
             if (!o.getClass().isInstance(value) && !(o instanceof Number && value instanceof Number)) {
                 MatchType.logger.warn(
-                    String.format("Incompatible type %s, %s",
-                        o.getClass().getCanonicalName(), value.getClass().getCanonicalName())
+                    "Incompatible type: Attribute value Type {}, Condition value Type {}",
+                    o.getClass().getCanonicalName(), value.getClass().getCanonicalName()
```

Comment:
We should either:
* turn _this_ into the place where we log
  ```
    "Audience condition \"{}\" evaluated to UNKNOWN because a value of type \"{}\" was passed for user attribute \"{}\""
  ```
  , which would require passing down information about the audience condition,
* or we should avoid logging a warning here, instead expecting `UserAttribute.evaluate` to [figure out and log what happened](https://github.com/optimizely/java-sdk/pull/251/files/5b35599ab7905158a3328d659d79da5b5f02e920#diff-289dc99f3875d89845c39fb04ac6b9c1R92).

@loganlinn any thoughts? And my apologies, I might be misunderstanding your original point or missing what's going on with this code.

Reply:
I'm starting to lean towards the second approach. It's definitely hacky (since the code that does the logging is attempting to figure out what went wrong in the code that evaluates the audience condition) but I like all the [customer-friendly log messages](https://github.com/optimizely/java-sdk/pull/251/files/5957ef142dcc245090c466b2dcd1f384f5584b5e#diff-289dc99f3875d89845c39fb04ac6b9c1R92) that we now have in `UserAttribute.evaluate`

Reply:
As of [commit 2be5a31](https://github.com/optimizely/java-sdk/pull/251/commits/2be5a3104b2e5b171f9ca07f353aa6366f514c5a), it looks like we're going with the second approach. Thanks for the updates! I'll resolve this conversation.